### PR TITLE
Fixes converted ads keep showing

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/8645

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

Please see https://github.com/brave/brave-browser/issues/8645

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: vendor/bat-native-ads/src/bat/ads/internal/ad_conversions.cc [hunk 1]
```diff
@@ -69,7 +90,117 @@ void AdConversions::ProcessQueue() {
...
+void AdConversions::OnGetAdConversions(
+    const std::string& url,
+    const Result result,
+    const AdConversionList& ad_conversions) {
+  if (result != SUCCESS) {
+    BLOG(ERROR) << "Failed to check ad conversions";
+    return;
+  }
+
+  std::deque<AdHistory> ads_history = client_->GetAdsShownHistory();
```
Comment (moritzhaller) [inline vendor/bat-native-ads/src/bat/ads/internal/ad_conversions.cc]: maybe `using` type, e.g. `AdHistoryList`

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ad_conversions.cc]: A larger change throughout codebase and as we may uplift want to reduce changes. Thanks

File: vendor/bat-native-ads/include/bat/ads/ads_history.h [hunk 2]
```diff
@@ -23,7 +23,8 @@ struct ADS_EXPORT AdsHistory {
 
   enum class FilterType {
     kNone = 0,
-    kConfirmationType
+    kConfirmationType,
```
Comment (masparrow) [inline vendor/bat-native-ads/include/bat/ads/ads_history.h]: Question only: can we/should we change `kConfirmationType` to be more specific to it's use, given the new `kAdConversionConfirmationType` below?

Comment (tmancey) [reply vendor/bat-native-ads/include/bat/ads/ads_history.h]: Agreed so will raise another ticket as want to keep this PR bare minimum as may be uplifted

File: vendor/bat-native-ads/src/bat/ads/internal/ad_conversions.cc [hunk 3]
```diff
@@ -69,7 +90,117 @@ void AdConversions::ProcessQueue() {
...
+  ads_history = FilterAdsHistory(ads_history);
+  ads_history = SortAdsHistory(ads_history);
+
+  AdConversionList new_ad_conversions = ad_conversions;
+  new_ad_conversions = FilterAdConversions(url, new_ad_conversions);
```
Comment (moritzhaller) [inline vendor/bat-native-ads/src/bat/ads/internal/ad_conversions.cc]: would it be worthwhile returning on `new_ad_conversions.empty()`? Might be marginal..

File: vendor/bat-native-ads/src/bat/ads/internal/ad_conversions.cc [hunk 4]
```diff
@@ -69,7 +90,117 @@ void AdConversions::ProcessQueue() {
...
+      if (observation_window > time) {
+        // Observation window has expired
+        continue;
+      }
+
+      BLOG(INFO) << "Ad conversion for " << url << " with "
+          << ad_conversion.creative_set_id << " creative set id for "
+              << std::string(ad_conversion.type);
+
+      AddItemToQueue(ad.ad_content.creative_instance_id,
```
Comment (moritzhaller) [inline vendor/bat-native-ads/src/bat/ads/internal/ad_conversions.cc]: is this synchronous? I.e. if we have an ad with the same creative_set_id as the ad that just converted will it return in L137? Is this also the reason why we can't pull L133 outside the loop?

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ad_conversions.cc]: Yes to both, thanks

File: vendor/bat-native-ads/src/bat/ads/internal/ad_conversions.cc (hunk 3 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ad_conversions.cc]: Very marginal, so not sure worth it at this time. Thanks

File: vendor/bat-native-ads/src/bat/ads/internal/sorts/ad_conversions_descending_sort.cc [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+AdConversionsDescendingSort::~AdConversionsDescendingSort() = default;
+
+AdConversionList AdConversionsDescendingSort::Apply(
+    const AdConversionList& list) const {
+  auto sorted_list = list;
+
+  std::sort(sorted_list.begin(), sorted_list.end(),
+      [](const AdConversionInfo& a, const AdConversionInfo& b) {
+    return a.type == "postclick" && b.type == "postview";
```
Comment (masparrow) [inline vendor/bat-native-ads/src/bat/ads/internal/sorts/ad_conversions_descending_sort.cc]: nit: can we const these hard coded strings?

**Pull request merged**